File: src/main/scala/edu/knoldus/introduction/rdd/BaseRDD.scala
package edu.knoldus.introduction.rdd

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}


object StarterTry extends App {


  val conf = new SparkConf().setAppName("BigApple").setMaster("local")
  val sc = new SparkContext(conf)

  // Parallelized Collections

  val data = Array(1, 2, 3, 4, 5, 4)
  val distData = sc.parallelize(data)

  // Working with K-Vs
  val colOfTuples = distData.map((x) => (x, 1))

  val kk = colOfTuples.reduceByKey((a, b) => a + b)

  kk foreach (println)



  println(distData.reduce((a, b) => a + b))

  //closure case
  var counter = 10
  distData.foreach(a => counter = a + counter)
  println(s" The closure example result = $counter")


  //Creating RDD from a file

  val distFile = sc.textFile("src/main/resources/compressed.gz")
  val linesRDD = distFile.map(x => (x.length))
  val yaSumOfLines = linesRDD.reduce((a, b) => a + b)
  val sumOfLines = linesRDD.reduce(FunctionHolder.addFunc)
  println(sumOfLines, yaSumOfLines)

  object FunctionHolder {
    val addFunc = (a: Int, b: Int) => a + b
  }

}


class MyClass {
  def func1(s: String): String = "Hello"

  def doStuff(rdd: RDD[String]): RDD[String] = {
    rdd.map(func1)
  }
}

File: src/main/java/edu/knoldus/introduction/rdd/BaseRdd.java
package edu.knoldus.introduction.rdd;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.storage.StorageLevel;

import java.util.Arrays;
import java.util.List;

public class BaseRdd {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("Big Apple").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        List<Integer> data = Arrays.asList(1, 2, 3, 4, 5);
        JavaRDD<Integer> distData = sc.parallelize(data,10);

        distData.foreach(x->System.out.println(x));

        Integer reduce = distData.reduce((a, b) -> a + b);
        System.out.println(reduce);

        // Creating RDD from a file
        // =========================
        // TODO Try with different kind of files in the resources folder
        JavaRDD<String> lines = sc.textFile("src/main/resources/compressed.gz");
        JavaRDD<Integer> lineLengths = lines.map(s -> s.length());

        // TODO Where would the persist go?
        //lineLengths.persist(StorageLevel.MEMORY_ONLY());
        int totalLength = lineLengths.reduce((a, b) -> a + b);
        System.out.println(totalLength);

    }
}
